drivers: eth_xmc4xxx: Add option to load MAC address from EEPROM

talih0 opened the pull request:
Adds an option to load the MAC address from an external EEPROM chip, for example from Microchip 25AA02E48 (atmel,at25) which come pre-programmed with a unique MAC.

The priority for setting the MAC address in the driver is: (1) local-mac-address from devicetree
(2) Eeprom address
(3) Random address

pdgendt reviewed (changes requested):
Why does an ethernet driver need to know where the mac address is coming from?

If you need this, you should implement/use the `set_config` from the ethernet driver API.

pdgendt commented on drivers/ethernet/eth_xmc4xxx.c:
```diff
@@ -792,6 +793,30 @@ static inline int eth_xmc4xxx_init_timestamp_control_reg(ETH_GLOBAL_TypeDef *reg
 	return 0;
 }
 
+static int eth_xmc4xxx_get_mac_address_eeprom(const struct device *dev)
+{
+#if DT_INST_NODE_HAS_PROP(0, mac_eeprom)
+	struct eth_xmc4xxx_data *dev_data = dev->data;
+	const struct device *eeprom_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0, mac_eeprom));
+	int ret;
+
+	if (!device_is_ready(eeprom_dev)) {
+		LOG_ERR("EEPROM device not ready.");
+		return -ENODEV;
+	}
+
+	ret = eeprom_read(eeprom_dev, CONFIG_ETH_XMC4XXX_MAC_EEPROM_ADDRESS, dev_data->mac_addr,
+			  sizeof(dev_data->mac_addr));
+	if (ret < 0) {
+		LOG_ERR("Error reading MAC address from EEPROM [%d]", ret);
+	}
+
+	return ret;
+#else
+	return -ENODEV;
+#endif
+}
```
Put the compile check around the function.

```c
#if DT_INST_NODE_HAS_PROP(0, mac_eeprom)
static int eth_xmc4xxx_get_mac_address_eeprom(const struct device *dev)
{
	struct eth_xmc4xxx_data *dev_data = dev->data;
	const struct device *eeprom_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0, mac_eeprom));
	int ret;

	if (!device_is_ready(eeprom_dev)) {
		LOG_ERR("EEPROM device not ready.");
		return -ENODEV;
	}

	ret = eeprom_read(eeprom_dev, CONFIG_ETH_XMC4XXX_MAC_EEPROM_ADDRESS, dev_data->mac_addr,
			  sizeof(dev_data->mac_addr));
	if (ret < 0) {
		LOG_ERR("Error reading MAC address from EEPROM [%d]", ret);
	}

	return ret;
}
#endif
```

talih0 replied:
In my patch returning -ENODEV makes using the function cleaner and the compiler optimizes out the function when the node is not in the devicetree.

pdgendt commented on drivers/ethernet/Kconfig.xmc4xxx:
```diff
@@ -46,6 +46,12 @@ config ETH_XMC4XXX_VLAN_HW_FILTER
 	  called multiple times, the filtering will be done on the latest
 	  tag.
 
+config ETH_XMC4XXX_MAC_EEPROM_ADDRESS
+	hex "EEPROM address of MAC"
+	default 0x0
+	help
+	  EEPROM address where the MAC address is stored.
```
We shouldn't be mixing Kconfig with device tree properties, it should be possible to have this added in the device tree:

```
mac-eeprom = <&eeprom 0x00>;
```

talih0 replied:
@pdgendt that would mean I'd have to modify the binding of the eeprom to add the -cells property?
@henrikbrixandersen would this be acceptable?

henrikbrixandersen replied:
If we are to do this (and not just rely on an application loading the board-specific MAC address from wherever and configuring the network interface), I'd prefer we align with how the Linux kernel devicetree bindings do this.

Something along the lines of:
- https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/nvmem/nvmem-consumer.yaml
- https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/aspeed/aspeed-bmc-asrock-e3c246d4i.dts (example use)

pdgendt replied:
> @pdgendt that would mean I'd have to modify the binding of the eeprom to add the -cells property?

I don't think it needs that see [the docs](https://docs.zephyrproject.org/latest/build/dts/phandles.html#zero-or-more-nodes-with-metadata-phandle-array-type)

talih0 replied:
@pdgendt, unless I'm missing something, it is the binding of the referenced node that describes how the specifier entries are parsed. (i.e. all the gpio and dma examples).

https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/devicetree/gpio.h#L69
From this example it's the gpio binding that needs to define gpio-cells.

henrikbrixandersen replied:
Correct. Please see the section on specifier spaces further down that page.

I'd still prefer using the same specifier space as the Linux kernel nvmem consumer bindings.

talih0 replied:
@henrikbrixandersen, I'll work on the nvmem consumer binding.

talih0 replied:
@henrikbrixandersen this an example from linux that matches what I'm trying to do:
![image](https://github.com/user-attachments/assets/a1c20ad6-14c9-4451-96f4-79503cdce17c)

But I have a devicetree question,  what would be the default compatible of mac_address in this case? I tried the same approach in zephyr. It works, but it doesn't say which compatible was used for this node:

![image](https://github.com/user-attachments/assets/410e773f-75a3-4b61-a095-c5d2993e62c3)

pdgendt reviewed (dismissed):
This should be built in CI, maybe add a case for `build_only` tests?

talih0 committed:
devicetree: Add nvmem-consumer

Adds property nvmem-cells for referencing a node that stores some
configuration information. A typical use case is the reading of MAC
address from an EEPROM device.

The referenced node is child of the non-volatile memory device. The
address of the referenced node stores address/offset where the
configuration is stored.

Th nvmem-cells phandle could be pointing to a child of an eeprom node
or an efuse register. Currently it is upto the devicetree driver to call
the correct API for reading the configuration data. Future work
could be to write an nvmem-producer driver abstraction that calls
correct api for flash/eeprom.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
drivers: eth_xmc4xxx: Use nvmem-cells for reading MAC from eeprom

Adds an option to load the MAC address from an external EEPROM chip,
for example from Microchip 25AA02E48 (atmel,at25) which come
pre-programmed with a unique MAC.

The priority for setting the MAC address in the driver is:
(1) local-mac-address from devicetree
(2) Eeprom address
(3) Random address

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
drivers: eth_sam_gmac: Use nvmem-consumer to read MAC address

Boards sam_e70_xplained, same54_xpro and sam_v71_xult are updated
to use nvem-consumer.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented:
When testing on the sam_v71_xult board I found that the ethernet did not work independent of my change. The read MAC address is the same before and after the change in this PR.

talih0 commented:
> When testing on the sam_v71_xult board I found that the ethernet did not work independent of my change. The read MAC address from the EEPROM chip is the same before and after the change in this PR.

I figured out the problem on the sam_v71_xult board. I was compiling for the wrong revision of the board: https://github.com/zephyrproject-rtos/zephyr/issues/77714

This PR was works fine on the sam_v71_xult/samv71q21b board

nandojve reviewed (approved):
Tested on sam_v71_xult/samv71q21 samples/net/cloud/tagoio_http_post.

talih0 commented:
@henrikbrixandersen, could you have another look, please?

talih0 commented:
@henrikbrixandersen, could you have another look when you have a moment please

henrikbrixandersen reviewed (changes requested):
Sorry about the long delay in reviewing, somehow this fell off my radar.

It's looking really, really good. I'd like to try getting this into Zephyr v4.2, although we're a bit pressed on time for this.

The tests for the macros can come in after the feature freeze, if needed.

pdgendt commented on include/zephyr/devicetree/nvmem.h:
```diff
@@ -0,0 +1,97 @@
+ *		};
+ *	};
+ *
+ *      eth: ethernet {
+ *		nvmem-cells = <&mac_address>;
+ *		nvmem-cell-names = "mac-address";
+ *	};
+ *
+ * Example usage:
+ *
+ *     DT_NVMEM_DEV_BY_IDX(DT_NODELABEL(eth), 0) // DEVICE_DT_GET(DT_NODELABEL(mac_eeprom))
```
Mixing tabs and spaces here

```suggestion
 *	eth: ethernet {
 *		nvmem-cells = <&mac_address>;
 *		nvmem-cell-names = "mac-address";
 *	};
 *
 * Example usage:
 *
 *	DT_NVMEM_DEV_BY_IDX(DT_NODELABEL(eth), 0) // DEVICE_DT_GET(DT_NODELABEL(mac_eeprom))
```

pdgendt commented on include/zephyr/devicetree/nvmem.h:
```diff
@@ -0,0 +1,97 @@
+ *		};
+ *	};
+ *
+ *     eth: ethernet@ {
+ *		...
+ *		nvmem-cells = <&mac_address>;
+ *		nvmem-cell-names = "mac-address";
+ *		...
+ *     };
+ *
+ * Example usage:
+ *
+ *     DT_NVMEM_ADDR_BY_IDX(DT_NODELABEL(eth), 0) // 0xfa
```
Same here

```suggestion
 *	eth: ethernet@ {
 *		...
 *		nvmem-cells = <&mac_address>;
 *		nvmem-cell-names = "mac-address";
 *		...
 *	};
 *
 * Example usage:
 *
 *	DT_NVMEM_ADDR_BY_IDX(DT_NODELABEL(eth), 0) // 0xfa
```

pdgendt commented on drivers/ethernet/eth_xmc4xxx.c:
```diff
@@ -799,6 +801,31 @@ static inline int eth_xmc4xxx_init_timestamp_control_reg(ETH_GLOBAL_TypeDef *reg
 	return 0;
 }
 
+static int eth_xmc4xxx_get_mac_address_eeprom(const struct device *dev)
```
Instead of putting this in the drivers, wouldn't make sense to put this on a higher level and use `eth_xmc4xxx_set_config` to apply?

maass-hamburg replied:
+ 1

maybe we can have some generic implementation for setting the mac address in 
in `net_if_post_init()` just before bringing the iface up of `subsys/net/ip/net_if.c`. 

What do you think? @jukkar @rlubos

maass-hamburg replied:
should be after iface init, but before iface up

talih0 commented:
I somehow missed the messages on this PR. Thank you to everyone for looking over. I'll send an updated version this week.

pdgendt commented:
> I somehow missed the messages on this PR. Thank you to everyone for looking over. I'll send an updated version this week.

@talih0 any update here?

maass-hamburg reviewed (changes requested):
should be implemented in a generic way, so that it uses the set_config api, so
that we don't have to duplicate code in every ethernet driver for that.

GseoC commented:
Hello,
We are also interested in the nvmem part. I'd like to use the `nvmem-cells` property to read the MAC address from fuses.
The thing is, with what is proposed here, this property is only used to get the phandle reference to the node. Therefore, it's the driver's job to call the appropriate API (EEPROM, OTP, ...). Wouldn't it be interesting  to have a NVMEM abstraction layer that redirects the read/write operations to the appropriate driver using a provider/consumer mechanism (like the YAML teases it)? I'm new to Zephyr so do not hesitate to point-out any constraints that may challenge this idea.

zagor commented:
I agree with others that this needs a more generic solution.

The mac address can be stored/obtained in a myriad different ways, only known by the application. Can we use an application-supplied callback that supplies the mac address? Then we don't have to implement every possible way addresses could be obtained.

maass-hamburg commented:
> The nvmem solution is nice, but I don't think it fully removes the need for a plain "use this mac address" approach.
> 
> The eeprom can be read protected or encrypted, or only store the last three bytes of the mac addres. The mac address could also be stored in a nand flash, or in a file on an emmc.
> 
> We can't solve every corner case in the device tree. We also need a way for the application code to simply provide the mac address.

There is already a api for that via set_config

zagor commented:
> There is already a api for that via set_config

Is there? I struggle to find how to do this early enough that no ethernet frames with the wrong mac address leak out of the interface.

pdgendt commented:
> The nvmem solution is nice, but I don't think it fully removes the need for a plain "use this mac address" approach.
> 
> The eeprom can be read protected or encrypted, or only store the last three bytes of the mac addres. The mac address could also be stored in a nand flash, or in a file on an emmc.

True, but it's a start.

> > There is already a api for that via set_config
> 
> Is there? I struggle to find how to do this early enough that no ethernet frames with the wrong mac address leak out of the interface.

Yeah, this is difficult indeed, for that reason I've created #96598 as an alternative here, but still do it in the driver init.

pdgendt commented:
Superseded by https://github.com/zephyrproject-rtos/zephyr/pull/96379 and https://github.com/zephyrproject-rtos/zephyr/pull/96598
